### PR TITLE
ENH: Replace elastix ImageSpatialObject2 by ITK ImageSpatialObject

### PR DESCRIPTION
**Pull request by N-Dekker**
Replaced elastix specific `ImageSpatialObject2` and `ImageMaskSpatialObject2` by ITK's `ImageSpatialObject` and `ImageMaskSpatialObject`, respectively.

Let azure CI build with the ITK5 version of May 9, 2019 (from ITK git master branch).

Added Mask->Update() calls to `MultiResolutionRegistration::UpdateMasks`, to ensure that the bounding boxes are computed.

Part of the elastix upgrade to ITKv5.

### CODE REVIEW

**Comment by N-Dekker** [comment]
Everything compiled well on all platforms. Only some CTest Elastix test failures on  Windows ITKv5.  https://dev.azure.com/kaspermarstal/Elastix/_build/results?buildId=414

Could this be just a hick up of the machine? I'll try to trigger a rebuild.

**Comment by mstaring** [comment]
looks great!

Where you able to check that the registrations (with masks) return identical results?
That should be done before merging this PR.

**Comment by N-Dekker** [comment]
Just updated ITKv5 SHA at the CI system, to ensure that this commit is included: https://github.com/InsightSoftwareConsortium/ITK/commit/7b4717d6cf3effa6645c3574efce15f266c3d4c7 "BUG: Half a pixel margin ImageMaskSpatialObject::ComputeMyBoundingBox()". I certainly won't merge when any test still fails at https://dev.azure.com/kaspermarstal/Elastix !

**Comment by mstaring** [comment]
Also, I am not sure we test registrations with a mask, so please check that we actually check for this functionality.

**Comment by N-Dekker** [comment]
CTest Elastix test failures on Windows ITKv5:
	 91 - elastix_run_3DCT_lung.MI.bspline.SGD.004_OUTPUT (Failed)
	 92 - elastix_run_3DCT_lung.MI.bspline.SGD.004_COMPARE_OVERLAP (Failed)
	 93 - elastix_run_3DCT_lung.MI.bspline.SGD.004_COMPARE_LANDMARKS (Failed)
	 94 - elastix_run_3DCT_lung.MI.bspline.SGD.004_COMPARE_TP (Failed)

**Comment by N-Dekker** [comment]
Still not OK, after the last force-push:

1>        Start  19: elastix_run_example_OUTPUT
1> 19/158 Test  #19: elastix_run_example_OUTPUT ..............................................   Passed    5.02 sec
1>        Start  20: elastix_run_example_COMPARE_IM
1> 20/158 Test  #20: elastix_run_example_COMPARE_IM ..........................................***Failed    0.40 sec

**Comment by N-Dekker** [comment]
Closed and reopened, just to trigger rebuild

**Comment by N-Dekker** [comment]
So https://dev.azure.com/kaspermarstal/Elastix/_build/results?buildId=422 now says, for Windows ITKv5:

> The following tests FAILED:
> Errors while running CTest
> 	 92 - elastix_run_3DCT_lung.MI.bspline.SGD.004_COMPARE_OVERLAP (Failed)
> 	 94 - elastix_run_3DCT_lung.MI.bspline.SGD.004_COMPARE_TP (Failed)

Details:
   
test 92
        Start  92: elastix_run_3DCT_lung.MI.bspline.SGD.004_COMPARE_OVERLAP

92: Test command: C:\hostedtoolcache\windows\Python\3.6.8\x64\python.exe "D:/a/1/s/Testing/elx_compare_overlap.py" "-b" "D:/a/1/Elastix-build/Testing/TransformParameters_3DCT_lung.MI.bspline.SGD.004.txt" "-d" "D:/a/1/Elastix-build/Testing/elastix_run_3DCT_lung.MI.bspline.SGD.004" "-m" "D:/a/1/s/Testing/Data/3DCT_lung_followup_mask.mha" "-p" "D:/a/1/Elastix-build/bin/Release/elxComputeOverlap.exe"
92: Test timeout computed to be: 1500

test 94
        Start  94: elastix_run_3DCT_lung.MI.bspline.SGD.004_COMPARE_TP

94: Test command: D:\a\1\Elastix-build\bin\Release\elxTransformParametersCompare.exe "-base" "D:/a/1/Elastix-build/Testing/TransformParameters_3DCT_lung.MI.bspline.SGD.004.txt" "-test" "D:/a/1/Elastix-build/Testing/elastix_run_3DCT_lung.MI.bspline.SGD.004/TransformParameters.0.txt" "-a" "1e-3"
94: Test timeout computed to be: 1500
94: Baseline transform parameters: D:/a/1/Elastix-build/Testing/TransformParameters_3DCT_lung.MI.bspline.SGD.004.txt has 51300 parameters.
94: Test transform parameters:     D:/a/1/Elastix-build/Testing/elastix_run_3DCT_lung.MI.bspline.SGD.004/TransformParameters.0.txt has 51300 parameters.
94: The norm of the difference between baseline and test transform parameters was computed, using
94:     || baseline - test ||
94:     ---------------------
94:        || baseline ||
94: Computed difference: 72.3165 / 201.03 = 0.35973
94: Allowed  difference: 0.001
94: ERROR: The difference is larger than acceptable!

**Comment by N-Dekker** [comment]
It appears that the differences between test results of this pull request versus the develop branch (as seen in Testing\elastix_run_3DCT_lung.MI.bspline.SGD.004\elastix.log) are gone when the following `IsInside` check (which should be redundant) is removed from the develop branch:

https://github.com/SuperElastix/elastix/blob/0a87c9df6aa39cc7432d86d732f57b9e46706f32/Common/itkImageMaskSpatialObject2.hxx#L72-L75

**Comment by mstaring** [comment]
ok great, when with that gone, the results are identical, I have faith we have a correct implementation of the masks and bounding boxes now. Then we can go ahead with updating baselines and merging the new code.

**Comment by N-Dekker** [comment]
@mstaring Do you agree that this pull request (which removes the old elastix `ImageSpatialObject2` and `itkImageMaskSpatialObject2`) is ready to merge? It still has regression failures, but _only_ when using ITK4. Which is basically elastix issue #153 (but then coming from ITK4 `ImageMaskSpatialObject::IsInside`)

File: Common/GTesting/itkImageMaskSpatialObject2GTest.cxx [hunk 1]
```diff
@@ -16,6 +16,9 @@
  *
  *=========================================================================*/
 
+#include <itkConfigure.h> // For ITK_VERSION_MAJOR
+
+#if ITK_VERSION_MAJOR < 5
 
 // First include the header file to be tested:
 #include "itkImageMaskSpatialObject2.h"
```
Comment (mstaring) [inline Common/GTesting/itkImageMaskSpatialObject2GTest.cxx]: this does not work anymore if this file is deleted right?

**Comment by mstaring** [comment]
i have one comments, otherwise good to go

File: Common/GTesting/itkImageMaskSpatialObject2GTest.cxx (hunk 1 above)
Comment (N-Dekker) [reply Common/GTesting/itkImageMaskSpatialObject2GTest.cxx]: Good point @mstaring ! I think the entire ImageMaskSpatialObject2 test should be removed, unconditionally. It wasn't such an interesting test anyway. Just my first attempt to get GoogleTest working with elastix.

**Comment by N-Dekker** [comment]
@mstaring I would like to keep ITK4 on the dashboard until the upgrade to ITK5 is finalized. But now (with this pull request) ITK4 has two test failures:

    elastix_run_3DCT_lung.MI.bspline.SGD.004_COMPARE_OVERLAP
    elastix_run_3DCT_lung.MI.bspline.SGD.004_COMPARE_TP

These can probably be fixed by either using the previous baseline (TransformParameters_3DCT_lung.MI.bspline.SGD.004.txt.in) for ITK4, or by just excluding these particular tests, for ITK4. Do you have a suggestion?

**Comment by mstaring** [comment]
Just exclude the test, ITK4 is not worth the trouble anymore ;-)

**Pull request merged**